route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -96,26 +96,30 @@ namespace route_following_plugin
 
         double target_speed=findSpeedLimit(current_lanelet);   //get Speed Limit
 
-        //double total_maneuver_length = current_progress + mvr_duration_ * target_speed; // TODO STOPPING_ISSUE we can do a min of this line and the line below. 
-        double total_maneuver_length = wm_->routeTrackPos(shortest_path.back().centerline2d().back()).downtrack; // TODO STOPPING_ISSUE Maneuver extends to the end of the route???
+        double total_maneuver_length = current_progress + mvr_duration_ * target_speed; // TODO STOPPING_ISSUE we can do a min of this line and the line below. 
+        double route_length= wm_->routeTrackPos(shortest_path.back().centerline2d().back()).downtrack; // TODO STOPPING_ISSUE Maneuver extends to the end of the route???
+        total_maneuver_length = std::min(total_maneuver_length, route_length);
+
         bool approaching_route_end = false;
         double time_req_to_stop,stopping_dist;
-        time_req_to_stop = sqrt(2*target_speed/jerk_); // TODO STOPPING_ISSUE target_speed not related to route end
+        time_req_to_stop = sqrt(2*findSpeedLimit(shortest_path.back())/jerk_); // TODO STOPPING_ISSUE target_speed not related to route end
         stopping_dist = target_speed*time_req_to_stop - (0.167 * jerk_ * pow(time_req_to_stop,3));
-        if(total_maneuver_length - current_progress <= stopping_dist){
+        
+        if(route_length - current_progress <= stopping_dist){
             approaching_route_end = true;
         }
         while(current_progress < total_maneuver_length && !approaching_route_end)
         {
             ROS_DEBUG_STREAM("Lanlet: " << shortest_path[last_lanelet_index].id());
             auto p = shortest_path[last_lanelet_index].centerline2d().back();
             double end_dist = wm_->routeTrackPos(shortest_path[last_lanelet_index].centerline2d().back()).downtrack;
+            end_dist = std::min(end_dist, total_maneuver_length);
+
             double dist_diff = end_dist - current_progress;
-            time_req_to_stop = sqrt(2*target_speed/jerk_);
-            stopping_dist = target_speed*time_req_to_stop - (0.167 * jerk_ * pow(time_req_to_stop,3));
             
-            if(total_maneuver_length - end_dist <= stopping_dist){
-                end_dist -= stopping_dist; // TODO STOPPING_ISSUE this will cause you to stop early if the final lanelet is shorter than stopping distance
+            if(route_length - end_dist <= stopping_dist){
+                end_dist -= stopping_dist; // TODO STOPPING_ISSUE this will cause you to stop early if the final lanelet is shorter than stopping distance 
```

Comment:
This still allows for stopping before the end if end_distance does not equal route length but the delta is less than stopping distance. 
I think better logic would be
end_dist = route_length - stopping_distance